azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java

```diff
@@ -1774,144 +1785,178 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
+      // get the locked flows for the project, so that they can be locked again after upload
+      final List<String> lockedFlows = getLockedFlows(project);
+
+      final Map<String, ValidationReport> reports = this.projectManager
+          .uploadProject(project, archiveFile, type, user, props);
+
+      if (this.enableQuartz) {
+        this.scheduler.schedule(project, user.getUserId());
+      }
+
+      // reset locks for flows as needed
+      lockFlowsForProject(project, lockedFlows);
+
+      // remove schedule of renamed/deleted flows
+      removeScheduleOfDeletedFlows(project, this.scheduleManager, (schedule) -> {
+        logger.info(
+            "Removed schedule with id {} of renamed/deleted flow: {} from project: {}.",
+            schedule.getScheduleId(), schedule.getFlowName(), schedule.getProjectName());
+        this.projectManager.postProjectEvent(project, EventType.SCHEDULE, "azkaban",
+            "Schedule " + schedule.toString() + " has been removed.");
+      });
+
+      registerErrorsAndWarningsFromValidationReport(resp, ret, reports);
+    } catch (final Exception e) {
+      logger.info("Installation Failed.", e);
+      String error = e.getMessage();
+      if (error.length() > 512) {
+        error = error.substring(0, 512) + "<br>Too many errors to display.<br>";
+      }
+      registerError(ret, "Installation Failed.<br>" + error, resp,
+          HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    } finally {
+      if (out != null) {
         out.close();
+      }
+      if (tempDir.exists()) {
+        FileUtils.deleteDirectory(tempDir);
+      }
+    }
 
-        if (this.enableQuartz) {
-          //todo chengren311: should maintain atomicity,
-          // e.g, if uploadProject fails, associated schedule shouldn't be added.
-          this.scheduler.unschedule(project);
-        }
+    logger.info("Upload: project " + projectName + " version is " + project.getVersion()
+        + ", reference is " + System.identityHashCode(project));
+    ret.put("version", String.valueOf(project.getVersion()));
+  }
 
-        // get the locked flows for the project, so that they can be locked again after upload
-        List<String> lockedFlows = getLockedFlows(project);
+  /**
+   * @return project. Null if invalid upload params or not enough permissions to proceed.
+   */
+  private Project validateUploadAndGetProject(final HttpServletResponse resp,
+      final Map<String, String> ret, final User user, final String projectName) {
+    if (projectName == null || projectName.isEmpty()) {
+      registerError(ret, "No project name found.", resp, HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    final Project project = this.projectManager.getProject(projectName);
+    if (project == null || !project.isActive()) {
+      final String failureCause = (project == null) ? "doesn't exist." : "was already removed.";
+      registerError(ret, "Installation Failed. Project '" + projectName + " "
```

Comment:
Missing the closing single-quota. Not worth its own PR, but if we are ever changing this, or anything close to it...